File: reg_schema/src/reg_schema/project_data.py
"""project_data.json schema models (see DESIGN.md → Two layers: models vs. validator).

Pure shape definitions, not validators. Pydantic v2 ``BaseModel`` —
``reg_schema`` is the deliberate exception to the workspace no-Pydantic
rule (``CLAUDE.md`` stack §): these models are the canonical project_data
shape, FastAPI response models in ``reg_webapp``, and the source of the
SPA's TypeScript types via ``model_json_schema()``.

Models are ``frozen`` + tuple-backed so consumers (reg_webapp and the SPA
via TS codegen) can hash, share, and pass instances freely. Pydantic
coerces list → tuple for ``tuple[...]`` fields automatically, so callers
may construct from list-shaped composites without losing the frozen +
hashable contract.

JSON deserialization and structural validation are deliberately
separate concerns:

- Structural rules (see DESIGN.md → Structural rules and issue codes — type/subtype consistency, FQID
  well-formedness, panel ordering, period grammar, etc.) live in
  ``validate_structural()`` and run on the **raw dict** before any
  model is constructed. They accumulate every issue into a
  ``ValidationResult`` rather than raising, so every runtime that shares
  the contract (SPA via TS mirror, webapp via direct import) sees the full
  issue list. The models intentionally do NOT re-encode those rules as
  raising field validators — that would replace the issue-accumulating
  contract with a fail-fast one.
- Models are constructed at boundaries (API ingress) from data that
  already passed ``validate_structural``. A Pydantic raise at that point
  signals validator/model drift, not user error.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

# Top-level enums (see DESIGN.md → Two layers: models vs. validator). Mirrored at runtime by the structural
# validator using ``get_args`` — same drift-protection pattern as
# ``IssueLevel`` in ``validation.py``.
Steward = Literal["global", "ifau", "swecov"]
ColumnType = Literal["id", "categorical", "numeric", "date", "datetime", "opaque"]
IdSubtype = Literal["integer", "string"]
NumericSubtype = Literal["integer", "double"]


class _Model(BaseModel):
    """Shared config: frozen (immutable + hashable) and extra-forbidding.

    ``extra="forbid"`` makes a typo in a constructed model fail loudly
    instead of dropping into defaults — the same drift guard the IR
    models use (``reg_meta_build.ir``).
    """

    model_config = ConfigDict(frozen=True, extra="forbid")


# Period -------------------------------------------------------------


class PeriodRange(_Model):
    """The ``{"from": ..., "to": ...}`` range form of ``Source.period``.

    Endpoints follow the same int / period-token-string forms as a bare
    period. ``from`` is a Python keyword, so the field is ``from_`` with a
    ``"from"`` alias; ``populate_by_name`` lets callers use either.

    This bare object is **only** legal as a ``Source.period`` value; a
    ``TimePoint`` range uses the discriminated ``TimeRange`` wrapper
    (``{"range": {...}}``) so ``TimeKey``'s union stays unambiguous.

    ``serialize_by_alias=True`` so ``model_dump()`` emits ``"from"`` (not the
    Python-safe ``"from_"``) without every caller having to pass
    ``by_alias=True`` — the un-aliased key would fail re-validation
    (``_is_period_range_obj`` requires exactly ``{"from", "to"}``).
    """

    model_config = ConfigDict(
        frozen=True,
        extra="forbid",
        populate_by_name=True,
        serialize_by_alias=True,
    )

    from_: int | str = Field(alias="from")
    to: int | str


# One contiguous piece of a ``Source.period``: bare year, period-token string,
# or explicit range. The ``"_default"`` snapshot sentinel is a plain string at
# the top level only — it is NOT a legal list member (structural rule).
PeriodSegment = int | str | PeriodRange

# ``Source.period``: a single segment, the ``"_default"`` sentinel (rides the
# ``str`` arm), or a LIST of segments — an interrupted series (#307, e.g.
# ``[{"from": 2005, "to": 2010}, {"from": 2015, "to": 2020}]``). The list form
# keeps one source = one register extraction (panel keys / binding sets are not
# duplicated across pseudo-sources). Structural rules for the list: non-empty,
# members are segments (no ``_default``, no nesting), sorted ascending and
# non-overlapping (adjacency allowed — the wire form stays canonical). Always
# required.
Period = PeriodSegment | tuple[PeriodSegment, ...]


# Binding ------------------------------------------------------------


class Binding(_Model):
    """A binding on a Source — one variable to include in the extract.

    ``variable`` is the binding FQID: ``<provider>/<register>/<slug>`` (3
    segments, see reg_meta/DESIGN.md → FQID grammar). Its ``provider/register`` prefix (first 2 segments) must
    equal the source's ``register_variant`` prefix — the variant is NOT
    repeated here, it lives once on the Source. That cross-field rule
    is enforced by the structural validator. There is no ``@version`` pin —
    that grammar is retired.

    A FQID names one CONCEPT. The reg_meta build enforces one value set per
    ``(variable, variant, period, delivery_column)``, but a concept may carry
    several co-existing delivery columns — parallel REPRESENTATIONS of it (SSYK
    3/4/5-digit, age 5/10-yr brackets). ``representation`` selects which one (by
    its ``variable_alias.delivery_column_name``); it is required only when the
    concept resolves to >1 column at the source's ``(variant, period)`` — the
    semantic validator (see reg_webapp/DESIGN.md → Semantic validation (semantic.py)) flags an ambiguous binding that omits it, and the
    SPA offers a chooser. A single-representation concept leaves it ``None``.

    ``display_name`` is optional: when absent, reg_meta-backed consumers
    resolve the default from ``variable_alias.delivery_column_name`` for
    the binding's state at the source's ``(register_variant, period)``. A
    reg_meta-free consumer that materializes data artifacts must resolve
    the default itself before emitting them — it never carries an
    unresolved ``display_name`` into its output.
    """

    variable: str
    type: ColumnType
    display_name: str | None = None
    id_subtype: IdSubtype | None = None
    numeric_subtype: NumericSubtype | None = None
    date_format: str | None = None
    datetime_format: str | None = None
    value_set: str | None = None
    representation: str | None = None


class Source(_Model):
    """A data source / table in the spec.

    ``register_variant`` is the 3-part variant **coordinate**
    (``<provider>/<register>/<variant>``) — not an FQID kind (see reg_meta/DESIGN.md → FQID grammar), but
    the same 3-part grammar. ``period`` is always required and polymorphic
    (``Period``). Together ``(register_variant's variant, period)`` selects
    each binding variable's ``variable_state``. ``name`` is the internal
    source handle referenced by panel members; it is not an FQID.
    """

    name: str
    register_variant: str
    period: Period
    bindings: tuple[Binding, ...]


# Panel ---------------------------------------------------------------
#
# Type aliases:
#
#   EntityKey = string | string[]                          // always column refs
#   TimePoint = int | string | LiteralPeriod | TimeRange   // string is column ref
#   TimeKey   = TimePoint | TimePoint[]
#
# Bare strings in panel keys are *always* column refs against a source's
# binding ``display_name`` values; literal string-shaped periods (e.g.
# ``"2018-01"``, ``"HT2018"``) must use the ``LiteralPeriod`` object form,
# and ranges the ``TimeRange`` wrapper. Integer literals stay as plain ints.


class LiteralPeriod(_Model):
    """The ``{"period": int | string}`` time_key form.

    The only way to express a string-shaped literal period at the
    schema level. Disambiguates ``"2018"`` (column ref) from
    ``{"period": "2018-01"}`` (literal period).
    """

    period: int | str


class TimeRange(_Model):
    """The ``{"range": {"from": ..., "to": ...}}`` time_key form.

    The discriminated wrapper for a period range in ``TimeKey`` position —
    distinct from the bare ``{"from", "to"}`` object, which is legal only
    as a ``Source.period`` (``PeriodRange``). The wrapper keeps the
    ``TimePoint`` union unambiguous.
    """

    range: PeriodRange


TimePoint = int | str | LiteralPeriod | TimeRange
TimeKey = TimePoint | tuple[TimePoint, ...]
EntityKey = str | tuple[str, ...]


class PanelMember(_Model):
    """A member of a Panel.

    ``source`` is the source ``name`` (the panel layer joins on
    delivered-data column headers, not FQIDs). ``entity_key`` /
    ``time_key`` override panel-level defaults; when both panel and member
    leave a key unset, it is inherited from the member's variant's
    ``panel_template`` at kit/bundle-build time — the structural
    validator does not flag the absence (it has no reg_meta).
    """

    source: str
    entity_key: EntityKey | None = None
    time_key: TimeKey | None = None


class Panel(_Model):
    """A panel definition over sources.

    Members are stored uniformly as ``PanelMember``. The bare-string
    shorthand (a source name with panel-level key defaults) is normalized
    to ``PanelMember(source=<name>)`` by the ``members`` validator, so
    consumers never branch on ``str | PanelMember``. Source-collision (each
    source belongs to at most one panel) and composite ordering /
    homogeneity rules are enforced by the structural validator, not here.
    """

    panel_id: str
    members: tuple[PanelMember, ...]
    entity_key: EntityKey | None = None
    time_key: TimeKey | None = None
    comment: str | None = None

    @field_validator(
        "members",
        mode="before",
        json_schema_input_type=tuple[str | PanelMember, ...],
    )
    @classmethod
    def _normalize_member_shorthand(cls, value: object) -> object:
        # A bare-string member is the source name. Expand it to the
        # object form before per-element validation; dicts and already-built
        # PanelMember instances pass through untouched.
        if isinstance(value, (list, tuple)):
            return [{"source": m} if isinstance(m, str) else m for m in value]
        return value


# Study window --------------------------------------------------------


class StudyWindow(_Model):
    """The optional ``{"from": <year>, "to": <year>}`` project study window.

    The global "project window" the redesigned subject page defaults each
    page's period picker to (see issue #611 → Period model). Deliberately
    NOT the full ``Period`` / ``PeriodRange`` grammar: it is a plain
    year-int pair, matching the year-granular header slider. Per-page
    deviation (months/quarters/terms, interrupted segments) keeps the rich
    grammar via ``?period``; this window only seeds the default.

    ``from`` is a Python keyword, so the field is ``from_`` with a ``"from"``
    alias — mirroring ``PeriodRange``. ``serialize_by_alias=True`` so
    ``model_dump()`` emits ``"from"`` (not ``"from_"``) without callers
    passing ``by_alias=True``.

    Endpoints are plain ``int`` years (not the ``int | str`` period-token
    forms of ``PeriodRange``): the window is year-granular by design. The
    only invariant is ``to >= from`` — a same-year window (``from == to``)
    is valid; a window can't end before it starts.
    """

    model_config = ConfigDict(
        frozen=True,
        extra="forbid",
        populate_by_name=True,
        serialize_by_alias=True,
    )

    from_: int = Field(alias="from")
    to: int

    @model_validator(mode="after")
    def _check_order(self) -> StudyWindow:
        if self.to < self.from_:
            raise ValueError(
                f"study window 'to' ({self.to}) must be >= 'from' ({self.from_})"
            )
        return self


# Top-level shape -----------------------------------------------------


class ProjectData(_Model):
    """The top-level ``project_data.json`` shape.

    The root is closed like every nested ``_Model``: an unknown field is a
    model-construction error. API boundaries still run the accumulating
    structural validator first so callers receive one stable
    ``unexpected_field`` issue per unknown key rather than a fail-fast Pydantic
    error. A future extension needs an explicit modeled container; arbitrary
    namespaced root blocks are not part of the v1 contract.
    """

    schema_version: str
    steward: Steward
    reg_meta_version: str
    name: str
    sources: tuple[Source, ...]
    panels: tuple[Panel, ...] = ()
    # Optional global study window (see issue #611 → Period model). Absent =
    # full history; existing specs validate unchanged (additive surface).
    window: StudyWindow | None = None

File: reg_webapp/frontend/src/lib/project_data.ts
/**
 * Pure project_data.json model helpers (NO runes — unit-tested in isolation;
 * `project_data.test.ts`). The canonical ProjectData root is CLOSED. The SPA's
 * in-memory draft remains deliberately raw enough to hold a malformed upload:
 * FOCUSED authoring scope (A5.3c) edits the top-level fields + `sources[]`
 * (`register_variant`, `period`, `bindings[]`); known `panels[]` and invalid
 * unknown root keys ROUND-TRIP VERBATIM until backend validation reports them.
 * Raw retention is a diagnostic path, not a supported extension mechanism.
 *
 * This is NOT a structural validator — the backend is canonical (see
 * reg_webapp/DESIGN.md → Pydantic boundary). These
 * helpers only construct + immutably edit the shape the SPA posts to
 * `/api/project/validate` / `/order`. The field names mirror
 * `reg_schema/src/reg_schema/project_data.py` (see reg_schema/DESIGN.md → Two
 * layers: models vs. validator).
 *
 * OpenAPI codegen documents the closed canonical request model, but the SPA cannot
 * use that strict type for its in-memory draft because `/validate` must also accept
 * malformed specs to diagnose them. The hand-written draft type therefore keeps an
 * index signature solely to retain invalid uploaded keys until diagnostics are
 * produced.
 */

/** A binding on a source — one variable to include. Only the fields the
 * A5.3c surface touches are named; any other key (`id_subtype`, `date_format`, …)
 * survives via the index signature. */
export interface Binding {
  variable: string;
  type: string;
  display_name?: string | null;
  value_set?: string | null;
  // The delivery column selecting which REPRESENTATION of the concept to extract
  // (set only when the concept resolves to >1 column at the source period — the
  // chooser fills it; the backend semantic check flags an ambiguous binding that
  // omits it). The job the retired `@version` pin once did, keyed on the column.
  representation?: string | null;
  [key: string]: unknown;
}

/** One contiguous piece of a `Source.period`: a bare year, a period-token
 * string, or a `{from, to}` range object. */
export type PeriodSegment =
  | number
  | string
  | { from: number | string; to: number | string };

/** A `Source.period` value: a segment, the `"_default"` sentinel (rides the
 * string arm), or a LIST of segments — an interrupted series (#307; the
 * backend enforces non-empty, sorted ascending, non-overlapping). Kept loose —
 * the server is the canonical period validator. */
export type Period = PeriodSegment | PeriodSegment[];

/** The optional global study window (the "project window", #611 → Period model).
 * A plain year-int pair matching reg_schema's `StudyWindow` wire shape (#613:
 * `{from, to}` int years, `to >= from`). NOT the full `Period` grammar — the
 * window is year-granular by design; per-page deviation keeps the rich grammar.
 * Absent = full history (backward-compatible — existing specs serialize
 * unchanged). project_data isn't a response model, so this isn't codegen'd into
 * `api-types` — it's hand-authored here alongside the `ProjectData` shape. */
export interface StudyWindow {
  from: number;
  to: number;
}

/** A data source / table. Open: panel-referenced or future keys survive. */
export interface Source {
  name: string;
  register_variant: string;
  period: Period;
  bindings: Binding[];
  [key: string]: unknown;
}

/**
 * The top-level project_data.json draft. Canonically closed; the index signature
 * exists only so an invalid uploaded key is not normalized away before validation.
 */
export interface ProjectData {
  schema_version: string;
  steward: string;
  reg_meta_version: string;
  name: string;
  sources: Source[];
  // The optional global study window (#611). Additive: omitted when unset (the
  // serializer drops `undefined` keys), so a project with no window round-trips
  // and validates unchanged.
  window?: StudyWindow;
  [key: string]: unknown;
}

/** The Model A `schema_version` a NEW draft is seeded with (reg_schema 2.0.0). */
export const MODEL_A_SCHEMA_VERSION = "2.0.0";

/** The 6 ColumnType values (reg_schema `ColumnType` `Literal`). The
 * BindingEditor's type `<select>` + the type-conditional advanced-field gating key
 * off this. Hand-maintained — `ColumnType` isn't on the OpenAPI surface
 * (project_data isn't a response model), so codegen can't supply it; co-located
 * with the `Binding` type it enumerates. */
export const COLUMN_TYPES = [
  "id",
  "categorical",
  "numeric",
  "date",
  "datetime",
  "opaque",
] as const;

/** Seed for a new project (from `/api/context`): the canonical reg_meta release
 * tag (derive it from the deployment's bare package version with
 * `regMetaReleaseTag`) + the deployment's steward id. `name` and `sources`
 * start empty. */
export interface ProjectSeed {
  reg_meta_version: string;
  steward: string;
}

/** Construct a fresh Model A skeleton. The version fields are seeded from
 * the deployment context; `schema_version` 2.x is the Model A gate, while
 * `reg_meta_version` records the catalog release used by this deployment. */
export function newProjectData(seed: ProjectSeed): ProjectData {
  return {
    schema_version: MODEL_A_SCHEMA_VERSION,
    steward: seed.steward,
    reg_meta_version: seed.reg_meta_version,
    name: "",
    sources: [],
  };
}

/**
 * Format the deployment's bare reg_meta PACKAGE version
 * (`context.webapp.reg_meta_version`, e.g. `"1.0.0"`) into the canonical
 * project_data release-tag form (`"reg_meta/v1.0.0"`; see reg_meta/DESIGN.md →
 * Release tags and distribution). Empty in → empty out (the context hasn't
 * resolved yet; the seed is corrected on the next New).
 */
export function regMetaReleaseTag(packageVersion: string): string {
  return packageVersion ? `reg_meta/v${packageVersion}` : "";
}

// ── Type guards ─────────────────────────────────────────────────────────────

/** A STRICT plain-object guard: true only for a non-null, non-array object. The
 * shared form of the "is this a JSON object, not `null` and not an array" check that
 * the store's open-file guard and the editors' malformed-slot fallbacks all need
 * (a verbatim-loaded draft can carry a `null`/array where an object is expected). */
export function isPlainObject(
  value: unknown,
): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

/** A read-side source slot after the SPA's malformed-slot seam has run. A plain
 * object is safe to inspect field-by-field, but not assumed structurally valid;
 * `null` means the original `sources[]` slot was `null`, an array, or another
 * non-object value. The raw draft is still kept verbatim for serialize/validate. */
export type SafeSource = Record<string, unknown> | null;

export function asSafeSource(slot: unknown): SafeSource {
  return isPlainObject(slot) ? slot : null;
}

/** Read-side view of a draft's `sources`. Non-array `sources` renders as empty;
 * malformed array slots stay counted as `null` so `/sources/{i}` addressing and
 * degraded cards line up with backend validation paths. */
export function safeSourceSlots(sources: unknown): SafeSource[] {
  return Array.isArray(sources) ? sources.map(asSafeSource) : [];
}

export function safeSourceName(source: unknown): string {
  const safe = asSafeSource(source);
  return typeof safe?.name === "string" ? safe.name : "";
}

export function safeSourceRegisterVariant(source: unknown): string {
  const safe = asSafeSource(source);
  return typeof safe?.register_variant === "string"
    ? safe.register_variant
    : "";
}

export function safeSourcePeriod(source: unknown): Period | null {
  const safe = asSafeSource(source);
  return safe != null && "period" in safe ? (safe.period as Period) : null;
}

export function safeSourceBindings(source: unknown): Binding[] {
  const safe = asSafeSource(source);
  return Array.isArray(safe?.bindings) ? (safe.bindings as Binding[]) : [];
}

export function sourceBindingsMalformed(source: unknown): boolean {
  const safe = asSafeSource(source);
  return (
    safe != null && safe.bindings !== undefined && !Array.isArray(safe.bindings)
  );
}

// ── Immutable top-level edits ───────────────────────────────────────────────
// Every mutator returns a NEW object (shallow clone + replaced slice) so the
// store can swap the `$state` reference and `dirty` recomputes. Unmapped keys on
// the spread survive (the `...draft` carries known panels and raw invalid keys).

/** Replace a top-level scalar field (`name`, `steward`, `reg_meta_version`, …). */
export function updateField<K extends keyof ProjectData>(
  draft: ProjectData,
  key: K,
  value: ProjectData[K],
): ProjectData {
  return { ...draft, [key]: value };
}

// ── Immutable source edits ──────────────────────────────────────────────────

/** Coerce a draft's `sources` to an array. An opened spec may carry a malformed
 * non-array `sources` (kept verbatim for serialize/validate); these mutators match
 * the editors' coercion doctrine (SourceEditor/ProjectEditor render a non-array as
 * []) so a structural edit on such a draft starts from [] rather than spreading a
 * string into char "sources" or throwing on `.map`/`.filter`. The malformed value is
 * thus REPLACED by a well-formed array on the first structural edit (intentional —
 * the user is fixing the draft via the editor). */
function sourcesArray(draft: ProjectData): Source[] {
  return Array.isArray(draft.sources) ? draft.sources : [];
}

// ── Source-name prefill (#312) ──────────────────────────────────────────────
// Source names are panel-key join handles (reg_schema panels join on source
// name), so a prefill must be unique among the draft's sources. The prefill is
// advisory: it only ever replaces an empty name — never a user-entered name; the
// catalog add's create path (`newSource` in the store) is its single caller.

/** Default source name for a register_variant coordinate: the register slug
 * (segment 2 of `provider/register/variant`) uppercased — `scb/lisa/v1` →
 * `"LISA"` (Swedish register stubs are mostly acronyms). `""` when the
 * coordinate has no register segment yet. */
export function defaultSourceName(registerVariant: string): string {
  const slug = registerVariant.split("/")[1] ?? "";
  return slug.toUpperCase();
}

/** `base` if no OTHER source (case-sensitive, as the schema compares) already
 * uses it, else the first free `base_2`, `base_3`, … The source at
 * `excludeIndex` is ignored (it's the one being named). */
export function uniqueSourceName(
  sources: readonly unknown[],
  base: string,
  excludeIndex: number,
): string {
  const taken = new Set(
    sources.filter((_, i) => i !== excludeIndex).map(safeSourceName),
  );
  if (!taken.has(base)) {
    return base;
  }
  let n = 2;
  while (taken.has(`${base}_${n}`)) {
    n += 1;
  }
  return `${base}_${n}`;
}

/** Remove the source at `index` (no-op if out of range). */
export function removeSource(draft: ProjectData, index: number): ProjectData {
  return {
    ...draft,
    sources: sourcesArray(draft).filter((_, i) => i !== index),
  };
}

/** Patch the source at `index` with `patch` (shallow merge — preserves the
 * source's unmapped keys + its `bindings`). */
export function updateSource(
  draft: ProjectData,
  index: number,
  patch: Partial<Source>,
): ProjectData {
  return {
    ...draft,
    sources: sourcesArray(draft).map((s, i) => {
      if (i !== index || asSafeSource(s) == null) {
        return s;
      }
      return { ...s, ...patch };
    }),
  };
}

// ── Immutable binding edits ─────────────────────────────────────────────────

/** Remove the binding at `bindingIndex` from the source at `sourceIndex`. */
export function removeBinding(
  draft: ProjectData,
  sourceIndex: number,
  bindingIndex: number,
): ProjectData {
  return updateSourceBindings(draft, sourceIndex, (bindings) =>
    bindings.filter((_, i) => i !== bindingIndex),
  );
}

/** Internal: apply a transform to one source's `bindings` immutably. Coerces a
 * non-array `sources` / `bindings` to [] (the editors' doctrine) so a structural
 * binding edit never spreads a string or throws on `.map`. */
function updateSourceBindings(
  draft: ProjectData,
  sourceIndex: number,
  fn: (bindings: Binding[]) => Binding[],
): ProjectData {
  return {
    ...draft,
    sources: sourcesArray(draft).map((s, i) => {
      if (i !== sourceIndex || asSafeSource(s) == null) {
        return s;
      }
      return { ...s, bindings: fn(safeSourceBindings(s)) };
    }),
  };
}

/**
 * Serialize a draft to the wire JSON text the SPA posts / downloads. A stable,
 * pretty (2-space) JSON. Raw invalid keys are carried by `JSON.stringify` so an
 * opened malformed file remains diagnosable and is never silently repaired.
 * NOT key-sorted: insertion order is preserved (a re-serialized file is
 * structurally faithful to what was opened), which is also what the `dirty`
 * baseline compares against.
 */
export function serializeProjectData(draft: ProjectData): string {
  return JSON.stringify(draft, null, 2);
}
